### PR TITLE
Allow hidden units connections in GRBM

### PR DESCRIPTION
**Pull request by VolodyaCO**
Somehow, intrahidden connections were disallowed, even though all math allows handling connected hidden units. This PR gets rid of the blocker and adds a test for the quasi objective gradient computation in the case of a GRBM with hidden-hidden connections.

### CODE REVIEW

**Review by kevinchern — CHANGES_REQUESTED** [review]
Took a quick look, here are some changes:
1. Fix failing tests
2. Instead of removing the check, move it where relevant, i.e., when the assumption is required for correctness.
3. Add tests to catch the above^

File: releasenotes/notes/allow-connected-hidden-connections-5e829224cedcd2ea.yaml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+---
+features:
+  - Allow hidden units to be connected during GRBM construction.
```
Comment (kevinchern) [inline releasenotes/notes/allow-connected-hidden-connections-5e829224cedcd2ea.yaml]: use proper name and syntax (backticks, no abbreviation)

Comment (VolodyaCO) [reply releasenotes/notes/allow-connected-hidden-connections-5e829224cedcd2ea.yaml]: Done

**Comment by VolodyaCO** [comment]
> Took a quick look, here are some changes:
> 
> 1. Fix failing tests
> 2. Instead of removing the check, move it where relevant, i.e., when the assumption is required for correctness.
> 3. Add tests to catch the above^

@kevinchern as per 2, the check is already in place where relevant.

**Comment by kevinchern** [comment]
> > Took a quick look, here are some changes:
> > 
> > 1. Fix failing tests
> > 2. Instead of removing the check, move it where relevant, i.e., when the assumption is required for correctness.
> > 3. Add tests to catch the above^
> 
> @kevinchern as per 2, the check is already in place where relevant.

@VolodyaCO do you have unpushed changes? O_O

edit: my mistake; the current implementation already checks for disconnected hidden units when using `kind="exact-disc"`

File: tests/test_boltzmann_machine.py [hunk 2]
```diff
@@ -407,6 +406,64 @@ def test_quasi_objective_gradient_hidden_units(self):
...
+        v3_cond = torch.tensor(1.0).float()
+
+        # t = sufficient statistics = (v1, v2, v3, v1v2, v1v3, v2v3)
+        t_cond = torch.tensor([
+            v1_cond,
+            v2_cond,
+            v3_cond,
+            v1_cond * v2_cond,
+            v1_cond * v3_cond,
+            v2_cond * v3_cond,
```
Comment (kevinchern) [inline tests/test_boltzmann_machine.py]: This test is incorrect but the tests are passing.

The test is incorrect because the average of sufficient statistics != sufficient statistics of average spins.

`avg(v1v2) = ((-1*-1) + (-1*1) + (1*-1))/3 = (1-1-1)/3 = - 1/3`
`avg(v1)avg(v2) = (-1 -1 + 1) / 3 * (- 1 + 1 - 1) / 3 = -1/3 * (-1/3) = 1/9`

**Review by kevinchern — CHANGES_REQUESTED** [review]
I think there's another assumption of disconnected hidden units somewhere or it's time for me to sign off work 😆. Let's do a quick zoom chat

File: tests/test_boltzmann_machine.py (hunk 2 above)
Comment (kevinchern) [reply tests/test_boltzmann_machine.py]: **Zoomed: the issue was the plugin is [averaging spin values before computing statistics](https://github.com/dwavesystems/dwave-pytorch-plugin/blob/main/dwave/plugins/torch/models/boltzmann_machine.py#L534-L537)**

File: dwave/plugins/torch/models/boltzmann_machine.py [hunk 3]
```diff
@@ -429,7 +425,7 @@ def quasi_objective(
                 raise ValueError(
                     f"`kind` {kind} should not be specified if the model is fully visible.")
         return (
-            self.sufficient_statistics(obs).mean(0, True)
+            reduce(self.sufficient_statistics(obs), "... n -> n", "mean")
```
Comment (kevinchern) [inline dwave/plugins/torch/models/boltzmann_machine.py:428]: cool 🤩!!

**Review by kevinchern — APPROVED** [review]
LGTM! Thank you!!

**Review by thisac — APPROVED** [review]
LGTM

**Comment by VolodyaCO** [comment]
Reviewed this again and I will merge.
